shared/js/background/components/resource-loader.js

```diff
@@ -0,0 +1,198 @@
+import Dexie from 'dexie'
+import bundledEtags from '../../../data/etags.json'
+import { alarms } from 'webextension-polyfill'
+/**
+ * @typedef {'tds' | 'surrogates' | 'config'} ResourceName
+ *
+ * @typedef {object} ResourceConfig
+ * @property {ResourceName} name
+ * @property {string} [remoteUrl]
+ * @property {string} [localUrl]
+ * @property {number} [updateIntervalMinutes]
+ * @property {'json'|'text'} [format]
+ *
+ * @typedef {(resourceName: ResourceName, etag: string, value: any) => void} OnUpdatedCallback
+ * @typedef {import('../settings.js')} Settings
+ */
+
+export default class ResourceLoader {
+    /** @type {Dexie?} */
+    static dbc = null
+
+    /**
+     * @param {ResourceConfig} config
+     * @param {{
+     *  settings: Settings
+     * }} opts
+     */
+    constructor (config, { settings }) {
+        this.settings = settings
+        this.name = config.name
+        this.remoteUrl = config.remoteUrl
+        this.localUrl = config.localUrl
+        this.updateIntervalMinutes = config.updateIntervalMinutes || 0
+        this.format = config.format || 'json'
+        this.data = null
+        /** @type {OnUpdatedCallback[]} */
+        this._onUpdatedListeners = []
```

Comment:
The code this was taken from had a comment suggesting we should use [EventTarget](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget) here instead (like we do in `shared/js/background/settings.js`). I think that still stands, why not change that while refactoring this code? _(If you prefer, we also have EventEmitter2 available, but I figure EventTarget is preferable as it's a built-in API.)_

If we did that, we could then also add a second emitter for `onAfterUpdated` which would fire after all the update listeners are done. We could then use that for our tests, to avoid the need for some of the backgroundWait.js hacks. _We might want to give that emitter a name to signify it's for tests only though, e.g. `_testOnlyOnAfterUpdated`._

Reply:
So I think `onConfigUpdate()` in `dnr-config-rulesets.js` returns a promise that resolves after the work updating the dynamic rules is done, so I think that yes the `onAfterUpdated` event could wait for all the `onUpdate` listeners to have been called and any returned Promises to have been resolved. It doesn't seem too hard, unless I'm missing something?